<a href="https://colab.research.google.com/github/BrettonSteiner/cse380-notebooks/blob/master/ponder_and_prove_combinatorics_and_probability.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Ponder and Prove Combinatorics and Probability
#### Due: Saturday, 6 February 2021, 11:59 pm.

## Conjecture

A number-theoretic conjecture of combinatorial significance is the following:

$degree2({2n \choose n}) =$ the "bits-on count" (or population count, or Hamming weight) of $n$.

$degree2(m)$ is defined as the number (degree, exponent) of 2's in the prime factorization of $m$.

In other words, for any $m$, a positive integer, $m = 2^e \cdot o$ where $o$ is an odd positive integer (could be 1) and $e$ is a natural number, including zero --- which would be the case when $m$ is odd. It's the $e$ that is the $degree2$ of $m$.

Another way to state this conjecture is that the number of 1's in the binary expansion of ${2n \choose n}$ for positive integer $n$ is equal to the number of 2's in the prime factorization of $n$.

Your task is to write Python code to test this conjecture for as many positive integers as you can. See the self-assessment for more details.

Note: a `bitsoncount` function can be a one-liner in Python: `return bin(x).count('1')`



In [31]:
from math import gcd

# Thanks Brother Neff!
def bits_on_count(x):
  return bin(x).count('1')

# Again, thank you Brother Neff!
def nCk(n, k):
  if k < 0 or k > n:
    return 0
  else:
    result = 1
    d = 1
    g = 1
    m = min(k, n - k)
    while (d <= m):
      g = gcd(result, d)
      result = n * (result // g)
      result = (result // (d // g))
      n -= 1
      d += 1
    return result

# This is a function derived from an example I found here:
# https://www.geeksforgeeks.org/count-occurrences-of-a-prime-number-in-the-prime-factorization-of-every-element-from-the-given-range/
def get_num_twos(n):
  prime = 2 # The prime we are looking for
  count = 0 
  val = prime
  while (True): 
    # Number of values in n that are divisible by val 
    a = n // val

    # Number of values in n - 1 that are divisible by val 
    b = (n - 1) // val

    # Increment the power of the val 
    val *= prime

    # (a - b) is the count of numbers in n that are divisible by val 
    if (a - b): 
      count += (a - b)

    # No values that are divisible by val 
    # thus exiting from the loop 
    else: 
      break

  return int(count); 

def test_conjecture(n):
  return bits_on_count(n) == get_num_twos(nCk(2*n,n))

In [34]:
import datetime

n = 1
try:
  while test_conjecture(n):
    n += 1

  # If it ends early due to a False statement:
  print('Verified up to ' + str(n) + ' at ' + str(datetime.datetime.now()))
  print('But then returned false')
except:
  print('Verified up to ' + str(n) + ' at ' + str(datetime.datetime.now()))

Verified up to 8116 at 2021-02-04 22:47:09.138715


## Basic Probability Theory Question
A dark room contains two barrels. The first barrel is filled with green marbles, the second is filled with a half-and-half mixture of green and blue marbles. So there's a 100% chance of choosing a green marble from the first barrel, and a 50% chance of choosing either color in the second barrel. You reach into one of the barrels (it's dark so you don't know which one) and select a marble at random. It's green. You select another. It's green too. You select a third, a fourth, a fifth, etc. Green each time. What is the *minimum* number of marbles you need to select to *exceed* a probability of 99% that you are picking them out of the all-green barrel? (Note that there are enough marbles so that the answer does not depend on how many marbles are in the second barrel.)


Note: These equations came from the Hint section on page 175 of the *First Three Odds* textbook.

The probability of choosing a **green** marble from the
- all-green barrel $= 1$
- half-and-half barrel $= (1/2)^n$

The probability of choosing a **blue** marble from the
- all-green barrel $= 0$
- half-and-half barrel $= 1-(1/2)^n$

I can use the equation for the probability of getting a blue marble from the half-and-half barrel to tell at what number of marbles I choose $n$ which I will have more than a 99% probability of getting a blue marble.

In [4]:
def green_marble_probability(n):
  return (1/2) ** n

def blue_marble_probability(n):
  return 1 - (1/2) ** n

for n in range(10):
  print(f'n: {n}')
  print(f'green: {green_marble_probability(n)}')
  print(f'blue: {blue_marble_probability(n)}')
  print()

n: 0
green: 1.0
blue: 0.0

n: 1
green: 0.5
blue: 0.5

n: 2
green: 0.25
blue: 0.75

n: 3
green: 0.125
blue: 0.875

n: 4
green: 0.0625
blue: 0.9375

n: 5
green: 0.03125
blue: 0.96875

n: 6
green: 0.015625
blue: 0.984375

n: 7
green: 0.0078125
blue: 0.9921875

n: 8
green: 0.00390625
blue: 0.99609375

n: 9
green: 0.001953125
blue: 0.998046875



I will have a probability greater than 99% that I am pulling marbles from the all-green barrel if I choose 7 or more green marbles consecutively  from the barrel.

## A Related But Deeper Basic Probability Theory Question
Take a deep breath. Suppose Shakespeare's account is accurate and Julius Caesar gasped "You too, Brutus" before breathing his last. What is the probability that you just inhaled a molecule that Julius Caesar exhaled in his dying breath?

Assume that after more than two thousand years the exhaled molecules are uniformly spread about the world and the vast majority are still free in the atmosphere. Assume further that there are $10^{44}$ molecules of air in the world, and that your inhaled quantity and Caesar's exhaled quantity were each about $2.2 \times 10^{22}$ molecules.
### Hint
If a number $x$ is small, then $(1 - x)$ is approximately equal to $e^{-x}$.


In [56]:
from math import e

# The amount of molecules on a single breath
breath_quantity = (2.2 * (10 ** 22))

print('The probability of a single molecule being from Julius Caesar\'s dying breath:')
pr_caesars_breath = breath_quantity / (10 ** 44)
print(pr_caesars_breath)
print()

print('The probability that no molecules in my breath are from Julius Caesar\'s dying breath:')
pr_none = e ** (-pr_caesars_breath * breath_quantity)
print(pr_none)
print()

print('The probability that at least one molecule in my breath is from Julius Caesar\'s dying breath:')
print(1 - pr_none)
print()

The probability of a single molecule being from Julius Caesar's dying breath:
2.1999999999999996e-22

The probability that no molecules in my breath are from Julius Caesar's dying breath:
0.00790705405159345

The probability that at least one molecule in my breath is from Julius Caesar's dying breath:
0.9920929459484066



The probability of inhaling a molecule of Julius Caeser's exhale from his dying breath is about 99.2% which is far more likely than I was expecting!

Daniel Strickland, Claire Hocker, and I used this reference to help us figure it out: https://puzzlemath.blogspot.com/2011/06/julius-caesars-last-breath.html

## What is True?
Assess yourself on how you did using the checkboxes below. Check a box by putting an 'X' in it only if it is warranted.


### What is true of my experience in general?
(5 points each, 15 points total)
- [ ] I had fun.
- [ ] I learned something new.
- [ ] I achieved something meaningful, or something I can build upon at a later time.

### What is true of my report on what I learned?
(5 points each, 25 points total)
- [ ] I wrote a sufficient number of well-written sentences.
- [ ] My report is free of "mechanical infelicities" (misspelled words, grammatical errors, punctuation errors, etc.).
- [ ] I reported on any connections I found between this investigation and something I already know.
- [ ] I reported who were and what contribution each of my collaborators made.
- [ ] I reported on how many numbers I was able to verify with a time/computation budget of 24 hours (in a row).


### What is true about my answers?
(15 points each, 60 points total)
- [ ] I figured out how to run a Python program continuously for at least 24 hours.
- [ ] I refrained from printing out anything except the highest number I verified, knowing that printing just slows a program down.
- [ ] I got the right answer for the first probability theory question.
- [ ] I got the right answer for the second probability theory question.
